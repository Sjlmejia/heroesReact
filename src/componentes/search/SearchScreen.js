import { useNavigate, useLocation } from 'react-router-dom';
import querString from 'query-string';


import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';
import { useForm } from '../hooks/useForm';
import { useMemo } from 'react';

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { q ='' } =querString.parse(location.search);
  const [state, handleInputChange]=useForm({
    searchText:q
  });
  const {searchText} =state;
  const heroes =useMemo(() => getHeroesByName(q), [q]) ;
  const handleSearch= (e) => {
    e.preventDefault();
    console.log(searchText)
    navigate(`?q=${ searchText }`)
  }
  return (
    <div>
      <h1>Busquedas</h1>
      <hr/>
      <div className="row">
        
        <div className="col-5">
          <h4>Buscar</h4>
          <hr/>

          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Buscar un heroe"
              className="form-control"
              autoComplete="off"
              name= "searchText"
              value={ searchText}
              onChange={handleInputChange}
            />
            <button className="btn btn-outline-primary mt-1" type="submit1">
              Buscar
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr />
          {
            (q === '')
              ?  <div className="alert alert-info">Buscar un heroe</div>
              : (heroes.length === 0) && <div className="alert alert-danger">No hay resultados: {q} </div>
          }
          {
            heroes.map(hero=>(
              <HeroCard 
                key={hero.id}
                {...hero}
              />
            ))
          }
        </div>
      </div>
    </div>
  )
}
