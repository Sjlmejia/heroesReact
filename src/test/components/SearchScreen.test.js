import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../componentes/search/SearchScreen";

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockNavigate,
}) );
describe('Pruebas en <SearchScreen />', () => { 
  test('debe de mostrar', () => { 
    const wrapper = mount(
    	<MemoryRouter initialEntries={['/search']}>
				<SearchScreen />
			</MemoryRouter>
    );
		expect(wrapper).toMatchSnapshot();
  });
	test('debe de mostrar Batman y el inpunt con el valor del query String', () => {
		const wrapper = mount(
    	<MemoryRouter initialEntries={['/search?q=batman']}>
				<SearchScreen />
			</MemoryRouter>
    );
		expect(wrapper.find('input').prop('value')).toBe('batman');
	});
	test('debe de mostrar un mensaje de error', () => {
		const wrapper = mount(
    	<MemoryRouter initialEntries={['/search?q=batman!']}>
				<SearchScreen />
			</MemoryRouter>
    );
		expect(wrapper.find('.alert').text().trim()).toBe('No hay resultados: batman!');
	})

	test('debe de llamar el navigate a la nueva pantalla', () => {
		const wrapper = mount(
    	<MemoryRouter initialEntries={['/search']}>
				<SearchScreen />
			</MemoryRouter>
    );

		const input = wrapper.find('input');
		input.simulate('change', {
			target: {
				name: 'searchText',
				value: 'batman'
			}
		})

		const form = wrapper.find('form');
		form.prop('onSubmit')({
			preventDefault(){}
		})

		expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
	})
})