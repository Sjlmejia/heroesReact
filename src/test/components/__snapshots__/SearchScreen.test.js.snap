// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar 1`] = `
<div>
  <h1>
    Busquedas
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit1"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un heroe
      </div>
    </div>
  </div>
</div>
`;
